File: src/year2025/day01.rs
type Input = Vec<i32>;

pub fn parse(input: &str) -> Input {
    input
        .lines()
        .filter_map(|l| {
           let line = l.trim();
           if line.is_empty() {
               return None;
           }

           let (direction, number_str) = line.split_at(1);
           let number: i32 = number_str.parse().ok()?;

           match direction {
               "R" => Some(number),
               "L" => Some(-number),
               _ => None,
           }
        }).collect()
}

pub fn part1(input: &Input) -> i32 {
    let mut point = 50;
    let mut password = 0;

    for &step in input {
        point = (point + step).rem_euclid(100);
        if point == 0 {
            password += 1;
        }
    }

    password
}

pub fn part2(input: &Input) -> i32 {
    let mut point = 50;
    let mut password = 0;
    let rhs = 100;

    for &step in input {
        if step >= 0 {
            password += (point + step) / rhs;
        } else {
            let reversed = (rhs - point) % rhs;
            password += (reversed - step) / rhs;
        }

       point = (point + step).rem_euclid(rhs);
    }

    password
}

File: go/year2025/day01/day01.go
package day01

import (
	"strconv"
	"strings"
)

type Input []int32

func Parse(input string) Input {
	var result Input

	for line := range strings.SplitSeq(input, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		if len(line) < 2 {
			continue
		}

		number, err := strconv.ParseInt(line[1:], 10, 32)
		if err != nil {
			continue
		}

		switch line[0] {
		case 'R':
			result = append(result, int32(number))
		case 'L':
			result = append(result, -int32(number))
		}
	}

	return result
}

func Part1(input *Input) int32 {
	point := int32(50)
	password := int32(0)

	for _, step := range *input {
		point = mod(point+step, 100)
		if point == 0 {
			password++
		}
	}

	return password
}

func Part2(input *Input) int32 {
	point := int32(50)
	password := int32(0)
	rhs := int32(100)

	for _, step := range *input {
		if step >= 0 {
			password += (point + step) / rhs
		} else {
			reversed := (rhs - point) % rhs
			password += (reversed - step) / rhs
		}

		point = mod(point+step, rhs)
	}

	return password
}

func mod(a, b int32) int32 {
	return ((a % b) + b) % b
}
